From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 16 Jul 2023 19:24:28 +0200
Subject: [PATCH] Client perspective - Determine and store visual
 representations

License: Fiddle Public License Agreement 1.0 (included in license/FPL-1.0.txt)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 0a46669503131fef418a4cd200061bc4d9c50495..1c63e1087313be59136d2e37a024fbf06d64a335 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -63,6 +63,7 @@ import net.minecraft.world.flag.FeatureFlags;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.LevelSettings;
 import net.minecraft.world.level.WorldDataConfiguration;
+import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.levelgen.WorldDimensions;
 import net.minecraft.world.level.levelgen.WorldOptions;
@@ -73,6 +74,8 @@ import net.minecraft.world.level.storage.LevelSummary;
 import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.WorldData;
 import org.bukkit.Material;
+import org.fiddlemc.fiddle.packet.ClientPerspectiveCategory;
+import org.fiddlemc.fiddle.packet.block.OnePerCategoryBlockStateReplacementProvider;
 import org.fiddlemc.fiddle.server.FiddleFeatureSupport;
 import org.fiddlemc.fiddle.configuration.FiddleConfigurations;
 import org.fiddlemc.fiddle.configuration.PermanenceConfirmation;
@@ -236,6 +239,19 @@ public class Main {
             Path path1 = Paths.get("server.properties");
             DedicatedServerSettings dedicatedserversettings = new DedicatedServerSettings(optionset); // CraftBukkit - CLI argument support
 
+            // Fiddle start - client perspective - determine and store visual representations
+            // TODO This doesn't do anything interesting yet: should be replaced with an actual mechanism that bases this on the resource pack assets
+            for (var block : BuiltInRegistries.BLOCK) {
+                if (block.isAddedByFiddle()) {
+                    for (var blockState : block.getStateDefinition().getPossibleStates()) {
+                        blockState.replacements = new OnePerCategoryBlockStateReplacementProvider(
+                            category -> category == ClientPerspectiveCategory.JAVA_WITH_FIDDLE_CLIENT_MOD ? blockState : Blocks.STONE.defaultBlockState()
+                        );
+                    }
+                }
+            }
+            // Fiddle end - client perspective - determine and store visual representations
+
             // Fiddle start - modifiable Bukkit enums - inject runtime versions - common utilities
             ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
             class InjectRuntimeClassVersionHelper {
